### PR TITLE
fix(testrunner): fix scenario where vstest would hang if no tests were detected

### PR DESCRIPTION
**Pull request by dupdob**
Fix issue #2549 : Stryker now warns when no test are present.
Fix issue #2559 : Stryker now disables parallelism for MsTest
Also implement a revised strategy for handling hanged VsTest:
1. For initial test session: wait for VsTest to stop, then wait for the end notification (if not received yet). No failover available for the time being.
2. For other test sessions: 
      1. wait for the end notification: if it times out, this session is discarded
      2. then, wait for the end of the session: if it times out, we assume VsTest hanged during clean up phase, it will be recycled, but the session is discarded.
      1. if anything timed out, VsTest console is recycled
      1. if the test session is discarded, a new attempt is made
      1. an exception is thrown after 5 failed attempts

### CODE REVIEW

**Comment by psfinaki** [comment]
Thanks for the fix! Maybe worth adding an integration test with an empty project?

**Comment by rouke-broersma** [comment]
Looks like it's a bit too aggressive, it marks the integration tests as having no tests and/or hanging.

image

**Comment by dupdob** [comment]
lol
sorry, checking that ASAP

**Comment by dupdob** [comment]
ok, got it. I introduced a hard coded 2 seconds timeout to the initial tests.
Fixing it

**Comment by dupdob** [comment]
should be ok now

File: src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs [hunk 1]
```diff
@@ -215,63 +239,70 @@ void HandlerUpdate(object sender, EventArgs e)
...
+                    _logger.LogError($"{RunnerId}: Test session ended with code {strykerVsTestHostLauncher.ErrorCode}");
                 }
-                _vsTestFailed = true;
-
             }
-            else if (strykerVsTestHostLauncher.ErrorCode > 0)
+            else
             {
-                _logger.LogError($"{RunnerId}: Test session ended with code {strykerVsTestHostLauncher.ErrorCode}");
+                // hard wait
```
Comment (psfinaki) [inline src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:291]: Are you sure this is a good idea? By far I don't have much understanding in the VSTest stuff, that's just paranoia from dev experience - having random problems because of hard waiting.

Maybe some very gracious timeout could be safer? Or, well, at least comment justifying the decision :)

Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:291]: hummm
I think I will comment. But I would gladly hear what a gracious timeout could be?
Note that nothing have actually changed here, I am merely restoring the historical behavior, that has been lost because I introduced a grace delay, hence transforming a null value to a 2 seconds timeout.

Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:291]: but it made me think of how the design should be improved

Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:291]: Well that probably depends on possible code flows that can lead here. I mean, I just wanted to share the fear that some other edge case project will fall here and things will hang forever. Not that I know what exactly.

What's the expected time to wait here, if you have any idea? For example, if usually it'd take a few seconds to wait, we could add a timeout of 5 mins and then throw something like "Hey, something weird is going on with the project X, report this on GitHub".

Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs]: yep, I reverted to the original implementation, which was delivering the expected results. I introduced changes as an attempt to work around the 'no test' problem, discovering later than nothing could be done. But I forgot to restore the original code.

**Comment by dupdob** [comment]
I revised the VsTest timeout logic again (a good night sleep does wonder) as I realized it was needed for some edge cases.
Indeed, Stryker will no longer attempts to run a test session without any known tests in it, but it may still happen for some other reason, such as some test filter.

On other news, I removed everything related to 'VsTest dit not' error messages. Previous assumptions was that no event whatsoever meant VsTest did not run/start properly, but it happens also when there is no test to run in the assembly.
So the message was misleading

**Comment by psfinaki** [comment]
The logic seems to be a bit more streamlined now, that's good :) 

But I still think an int test will be beneficial since this has broken a couple of times already.

**Comment by dupdob** [comment]
Update: PR switched to draft as I have a reproduceable error in solution mode.
Which is cool: another problem will be fixed!

File: src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestMockingHelper.cs [hunk 2]
```diff
@@ -130,19 +130,24 @@ protected SourceProjectInfo BuildSourceProjectInfo(IEnumerable<Mutant> mutants =
...
         => new(name, framework == TestFrameworks.xUnit ? _xUnitUri : _NUnitUri, _testAssemblyPath) { Id = new Guid(), DisplayName = displayName ?? name };
 
     private Microsoft.VisualStudio.TestPlatform.ObjectModel.TestCase FindOrBuildCase(string testResultId) => TestCases.FirstOrDefault(t => t.FullyQualifiedName == testResultId) ?? BuildCase(testResultId);
 
-    private static void MockTestRun(ITestRunEventsHandler testRunEvents, IReadOnlyList<TestResult> testResults,
+    private static Task MockTestRun(ITestRunEventsHandler testRunEvents, IReadOnlyList<TestResult> testResults,
         Microsoft.VisualStudio.TestPlatform.ObjectModel.TestCase timeOutTest = null) =>
         Task.Run(() =>
         {
+            if (testResults.Count == 0)
```
Comment (psfinaki) [inline src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestMockingHelper.cs:146]: Just thinking out loud: from a design perspective, this feels like a bit too much of white-box testing - I would expect this to be a transparent implementation detail in the code itself.

File: src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestRunnerPoolTests.cs [hunk 3]
```diff
@@ -156,6 +179,26 @@ public void DetectTimeout()
             result.TimedOutTests.IsEmpty.ShouldBeFalse();
         }
 
+        [Fact]
+        public void RetryHangedSession()
```
Comment (psfinaki) [inline src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestRunnerPoolTests.cs:183]: Okay sorry I cannot skip this anymore! _Hanged_ is used only for the method of execution, otherwise it's _hung_ :D

Comment (dupdob) [reply src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestRunnerPoolTests.cs:183]: yep, I did not remember the exact rule. Any useful synonym I could use instead (I am not fan of hung either)
:-)

File: src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestMockingHelper.cs (hunk 2 above)
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestMockingHelper.cs:146]: I do not think this is white box testing, in the sense that the test verifies what happens inside the _sut_.
Here it is about better faking for VsTest: if you launch VsTest with an empty test assembly, you will not get any event at all. While I suspect this is a bug on VsTest part, we still need to handle this gracefully

Don't you think?

Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestMockingHelper.cs:146]: Yeah fair, I didn't look closely enough. Okay then!

File: src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestRunnerPoolTests.cs (hunk 3 above)
Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestRunnerPoolTests.cs]: It could be _freeze_ probably, but I would live with whatever, just wanted to make a point :-P

File: src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs [hunk 4]
```diff
@@ -153,14 +153,13 @@ private ICollection<Guid> TestCases(IReadOnlyList<Mutant> mutants, Dictionary<in
...
-            bool compressAll = true)
+        private TestRunResult BuildTestRunResult(IRunResults testResults, int expectedTests, int totalCountOfTests, bool compressAll = true)
         {
             var resultAsArray = testResults.TestResults.ToArray();
             var testCases = resultAsArray.Select(t => t.TestCase.Id).ToHashSet();
             var ranTestsCount = testCases.Count;
-            var timeout = !_aborted && ranTestsCount < expectedTests;
-            var ranTests = compressAll && ranTestsCount >= totalCountOfTests ? (ITestGuids)TestGuidsList.EveryTest() : new WrappedGuidsEnumeration(testCases);
+            var timeout = !_cancelled && ranTestsCount < expectedTests;
+            var ranTests = compressAll && totalCountOfTests>0 && ranTestsCount >= totalCountOfTests ? (ITestGuids)TestGuidsList.EveryTest() : new WrappedGuidsEnumeration(testCases);
```
Comment (psfinaki) [inline src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:164]: So thinking out loud here - I guess at some point it could be useful to just write down all the known scenarios here, like:
- 1 source project x 1 test project
- 1 source project x many test projects
- many source projects x one test project
- many source projects x many test projects
- empty source projects
- empty test projects
- ... 

... and then put this in code (read - design patterns, interfaces, whatever). 

It seems like sooner or later in the flow most of these scenarios need special treatment hence I am under the impression that the related code is gradually turning into a bunch of if/else hacks and becoming a bit hard to follow.

My impression can be wrong, that's basically a general reminder to keep an eye on the design in the light of newly appearing edge cases and scenarious.

File: src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs [hunk 5]
```diff
@@ -195,13 +194,41 @@ private ICollection<Guid> TestCases(IReadOnlyList<Mutant> mutants, Dictionary<in
...
+            Action<IRunResults> updateHandler = null)
+        {
+
+            if (tests.IsEveryTest && !_context.IsValidSourceList(sources))
+            {
+                _logger.LogWarning($"Test assembl{(sources.Count() <2 ? "y does" : "ies do")} not contain any test, skipping.");
+                return new RunEventHandler(_context.VsTests, _logger, RunnerId);
+            }
+
+            for (var attempt = 0; attempt < 5; attempt++)
```
[the remark this answers is not in the document]
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs]: Done. 
There is no solid justifications, except VsTest having an history of time related bugs. But I think things get better, so I lowered the number of attempts to 3.

File: src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs (hunk 4 above)
Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:168]: Well, I put my comment here because these lines just feel like we are starting to take too many circumstances into account, but generally it's unrelated :) 

But yeah I guess I am overthinking here a bit, eventually we might also just end up with some kind of wrapper on top of VsTest which would effectively carry the semantics of "this is how VsTest would work if it was designed properly and implemented without bugs", basically cutting corners and streamlining edge cases.

Comment (dupdob) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:168]: Comment has been added.
VsTestRunner more or less acts as this VsTest wrapper. Maybe a better design will emerge or be found in the coming months. For my own part, I feel like I have a done a lot of redesign in the past year, and I want to focus on fixing things for the time being 😄

Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:168]: Sure, that's fine. I am just sharing thoughts here, no action required :D

File: azure-pipelines.yml [hunk 6]
```diff
@@ -197,6 +197,10 @@ stages:
       parameters:
         testName: 'windows-netframework'
         workingDirectory: 'integrationtest/TargetProjects/NetFramework/FullFrameworkApp.Test'
+    - script: $(Agent.BuildDirectory)/tools/dotnet-stryker
```
Comment (rouke-broersma) [inline azure-pipelines.yml:200]: I see the empty test project is dotnet6. The windows integration test is the slowest, it would be better to add this to the linux integration test phase as it's the quickest.

Comment (dupdob) [reply azure-pipelines.yml]: seems ok

**Comment by psfinaki** [comment]
Cool - thanks for merging and looking forward to the release to test this in the wild :)
